packages/graphql/src/services/practiceQuizzes.ts

```diff
@@ -27,32 +39,247 @@ export async function getPracticeQuizData(
+  const existingResponse = await ctx.prisma.questionResponse.findUnique({
+    where: {
+      participantId_elementInstanceId: {
+        participantId: ctx.user.sub,
+        elementInstanceId: id,
+      },
+    },
+  })
+
+  ctx.prisma.questionResponseDetail.create({
+    data: {
+      response: {
+        correctness: correctness,
+      },
+      participant: {
+        connect: { id: ctx.user.sub },
+      },
+      questionInstance: {
+        connect: { id },
+      },
+      participation: {
+        connect: {
+          courseId_participantId: {
+            courseId,
+            participantId: ctx.user.sub,
+          },
+        },
+      },
+    },
+  })
+
+  // TODO: think about possibility to fuse these two cases with a single upsert (however, keep in mind that we need to update JSON stuff here)
+  if (existingResponse) {
+    let aggregatedResponse =
+      existingResponse.aggregatedResponses as AggregatedResponseFlashcard
+    switch (correctness) {
+      case 0:
+        aggregatedResponse = {
+          ...aggregatedResponse,
+          wrong: aggregatedResponse.wrong + 1,
+          total: aggregatedResponse.total + 1,
+        }
+        break
+      case 1:
+        aggregatedResponse = {
+          ...aggregatedResponse,
+          partial: aggregatedResponse.partial + 1,
+          total: aggregatedResponse.total + 1,
+        }
+        break
+      case 2:
+        aggregatedResponse = {
+          ...aggregatedResponse,
+          correct: aggregatedResponse.correct + 1,
+          total: aggregatedResponse.total + 1,
+        }
+        break
+    }
+
+    // update existing response
+    const questionResponse = await ctx.prisma.questionResponse.update({
+      where: {
+        participantId_elementInstanceId: {
+          participantId: ctx.user.sub,
+          elementInstanceId: id,
+        },
+      },
+      data: {
+        trialsCount: {
+          increment: 1,
+        },
+        response: {
+          correctness,
+        },
+        aggregatedResponses: aggregatedResponse,
+        correctCount: {
+          increment: correctness === 2 ? 1 : 0,
+        },
+        correctCountStreak: {
+          increment:
+            correctness === 2 ? 1 : -existingResponse.correctCountStreak,
+        },
+        lastCorrectAt: correctness === 2 ? new Date() : undefined,
+        participant: {
+          connect: { id: ctx.user.sub },
+        },
+        elementInstance: {
+          connect: { id },
+        },
+        participation: {
+          connect: {
+            courseId_participantId: {
+              courseId,
+              participantId: ctx.user.sub,
+            },
+          },
+        },
+      },
+    })
+
+    return questionResponse
+  } else {
+    let aggregatedResponse = {
+      wrong: 0,
+      partial: 0,
+      correct: 0,
+      total: 0,
+    }
+
+    switch (correctness) {
+      case 0:
+        aggregatedResponse = {
+          wrong: 1,
```

Comment:
simplify by just changing the relevant numbers to 1 instead of redefining the entire object

Reply:
wip implementation has been committed

Reply:
successfully implemented